File: Data Structures & Algorithms/reorder-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        ArrayList<Integer> vals = new ArrayList<>();
        ListNode curr = head;

        while (curr != null) {
            vals.add(curr.val);
            curr = curr.next;
        }

        int i = 0; 
        int j = vals.size() - 1;
        boolean leftTurn = true;
        while (i <= j) {
            if (leftTurn) {
                head.val = vals.get(i);
                i++;
            }
            else {
                head.val = vals.get(j);
                j--;
            }
            leftTurn = !leftTurn;
            head = head.next;
        }

    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return 

        res = lists[0]
        for i in range(1, len(lists)):
            res = self.combine(res, lists[i])
        return res

    def combine(self, l1, l2):
        res = ListNode()
        cur = res
        a, b = l1, l2

        while a and b:
            if a.val < b.val:
                cur.next = ListNode(a.val)
                cur = cur.next
                a = a.next
            else:
                cur.next = ListNode(b.val)
                cur = cur.next
                b = b.next
        
        cur.next = a if a else b
        return res.next

        